### ISSUE TITLE
"Screen Reader detected" appears without using a screen reader

### ISSUE THREAD

**Issue by alexandrudima** [Microsoft/vscode#27893]
Extracted from @Tyriar in https://github.com/Microsoft/vscode/issues/27698#issuecomment-305585158

> I don't have a screen reader as far as I know on my Windows machine yet it's still there, removing my line/col information, bright green and in my face? As someone who has never used a screen reader on this machine I don't know how to disable the message.

**Comment by alexandrudima** [comment]
@Tyriar Do you perhaps use any software on this machine that behaves like a Screen Reader ? Do you do anything special that makes Electron enter accessibility mode or does Electron enter it on start-up? My best guess is that there must be some software on your machine that sends automation / accessibility events to VS Code, kicking Electron into entering accessibility mode.

Can you confirm that Chromium is in accessibility mode when this happens via the Help > Accessibility Options menu item. What software do you have running on this machine?

**Comment by alexandrudima** [comment]
@Tyriar are you 100% certain you didn't have NVDA running. e.g. I think this happened once to @bpasero where NVDA was running (and "hidden" in the task bar next to the clock), and without enabling sounds in a VM one would not even notice.

**Comment by alexandrudima** [comment]
This issue covers the false-positive case, while #27833 covers the false-negative case.

This is mostly based on a new setting `editor.accessibilitySupport`:
* 'auto' means that we will respect what Electron's API tells us <-- this is the default.
* 'off' means that we will never be in accessibility mode, irrespective of Electron's API (this is to cover the false-positive case).
* 'on' means that we will always be in accessibility mode, irrespective of Electron's API (this is to cover the false-negative case).


The scenario is:

### A screen reader is **NOT** attached to VS Code, but Electron's API tells us that a screen reader is attached
* If this ever happens in the wild, an Electron bug should be logged.
* The idea is we have a way to workaround this (a setting), such that you can turn off anything we do for screen readers if you are not using a screen reader.
* One posible way to get into this state is on Windows to start NVDA and then close NVDA. Electron does not let us know when a screen reader is detached and we remain in this mode.

Please check that:
* When the editor is in this mode, the label at the bottom helps you get out of this mode (i.e. it has a tooltip)
* once `editor.accessibilitySupport: "off"`:
   * even attaching a Screen Reader will not change anything (nothing in the status bar, nothing in the editor textarea value). However:
      * the editor aria label tells about `alt+f1`.
      * the `alt+f1` explains the current accessibility state thoroughly
      * there is an "emergency" way to turn on accessibility support via ctrl/cmd+e in case the setting was turned off by accident.

**Comment by Tyriar** [comment]
@alexandrudima 

> Do you perhaps use any software on this machine that behaves like a Screen Reader ?

OneNote and maybe [Dashlane](https://www.dashlane.com) would be the only things that comes to mind that may poke such APIs. This machine also has a touch screen and en, ja, zh and ko keyboard layouts active.

> Do you do anything special that makes Electron enter accessibility mode or does Electron enter it on start-up?

It happens on start up.

> are you 100% certain you didn't have NVDA running.

It's not in programs and features and I'm pretty sure I've never installed it

---

The tooltip and setting is fine for me, just hope that my case isn't too common.

**Comment by alexandrudima** [comment]
I hope so too. I have OneNote installed and I don't think that's it.

The interesting thing is that:
* the setting can stop us from doing anything accessibility related, but the setting does not stop Chromium.
* it is my experience that Chromium is *very slow* when it enters accessibility mode.

**Comment by Bill-Stewart** [comment]
I have the same problem. I do not have any screen reader software. I had to manually set `editor.accessibilitySupport` to `off`.

**Comment by Tyriar** [comment]
@Bill-Stewart can you think of any software you have that could be pretending to be a screen reader?

**Comment by Bill-Stewart** [comment]
No idea whatsoever.

**Comment by alexandrudima** [comment]
@Bill-Stewart Using Screen Readers on Windows ultimately result in a window receiving a very specific type of message from the OS, similar how a window receives for example a key down, or a mouse move, or a window resize.

I don't know all the details, but I believe a COM protocol based connection is then established between the Screen Reader and the window, and through this connection accessibility information travels.

Electron, through its use of Chromium, when it sees such a specific message type, decides to enable a flag that ultimately leads to data from the renderer process being sent over to the main process for accessibility purposes. We simply react when Electron lets us know it has entered this mode.

TL;DR: there might be software on your machine that is generating events going to windows that look very similar to the events that a Screen Reader would generate. Usually, UI automation software would use the same mechanisms as a Screen Reader. I'm not sure if it's useful or not, but if we'd get to know what software tricks Electron into entering accessibility mode, perhaps we can communicate that back to the Electron folks and they maybe could work around it.

**Comment by chrismbarr** [comment]
The same thing is happening to me.  I have no idea what software on my machine is thought to behave like a screen reader.

The strange part is that it's intermittent.  VSCode doesn't *always* launch with this green badge, but just sometimes.

This morining I got to my dek and opened VSCode and the badge was there: _(note the time)_
![image](https://user-images.githubusercontent.com/463685/28920083-4f64fee6-781e-11e7-89cc-17cb09bce996.png)

I saw it and search online, and found this issue.  As I was typing this I decided to close and re-launch VSCode - and now the badge is gone _(note the time again)_!

![image](https://user-images.githubusercontent.com/463685/28920179-b4f06598-781e-11e7-9acc-fe1b2e70c080.png)

The only thing I did while it was open is install a new extension.

**Comment by anthony-gregg** [comment]
I have an even better variation of this. I opened VS Code this morning and also had the "Screen Reader Detected" green badge and it also was missing the Line Number and Character location. I closed VS Code and reopened it and now look at it:

![image 22](https://user-images.githubusercontent.com/23386642/28939105-c2d0c3a6-7855-11e7-9edf-e75a38573c52.png)


I have not seen a scenario where both have been displayed based on the above statements. I also have no idea what software I have that is acting as a Screen Reader. I do have One Note, but that cannot be the case I would believe.

**Comment by alexandrudima** [comment]
@chrismbarr @anthony-gregg I'm sorry about you running into these false positive cases. Please change the setting `editor.accessibilitySupport` from `"auto"` to `"off"`. We need to continue to ship with `"auto"` for the users that are forced to use a Screen Reader.

When the setting is "auto", we react to Electron telling us a Screen Reader is attached and then we proceed to do some tweaks: e.g. disable word wrapping, as wrapping points cannot be communicated correctly to Screen Readers (w3c spec limitation) and visually impaired users would hear "New Line" at the location of a wrapping point, which is incorrect. Another example is our removal of the Ln/Col indicator which would produce accessibility events that would throw off / confuse / spam NVDA.

In any case, it is probably a good idea to find out what software on your machine causes this, as, e.g. from my experience, when Electron/Chromium enters accessibility mode, VS Code will run a lot slower for you. e.g. https://github.com/electron/electron/issues/7208 . It is also possible that Chrome on your machine will be running a lot slower than for other users. In Chrome, you can visit `chrome://accessibility/` to find out if Chrome has entered this mode on your machine.

**Comment by anthony-gregg** [comment]
Alexandru,

Thank you for your response and help. I have previously turned off my accessibility features in VS Code and my problem went away. I have also looked at my Chrome accessibility and found that it was not activated:

[cid:image003.jpg@01D31049.DD3DD3B0]


I am not sure what else could be causing the false reading.

Thanks

Anthony


Anthony Gregg | Technical Lead | Retail Development | Application Development & Maintenance | Best Buy | A5-071 | Ph.: (612) 291-2350 | Cell: (612) 231-0011   [cid:image001.jpg@01CCB595.50739B50] <http://www.bestbuy.com/>

TFS Requests:  Service Now <https://bestbuy.service-now.com/> – Self Service -> Request Catalog -> IT Infrastructure -> TFS Support Requests

**Comment by chrismbarr** [comment]
So here's something else. I was working on one project this morning and no screen reader badge appeared. Without closing that project, I opened up another one and that one **does** have the screen reader banner at the same time!

![image](https://user-images.githubusercontent.com/463685/29171700-5eea0414-7dab-11e7-85c9-050f64e146cb.png)


So, I now have an instance of VSCode in each mode.  Neat!

**Comment by anthony-gregg** [comment]
Very unique and I assume that you have Accessibility enabled, correct?


Anthony Gregg | Technical Lead | Retail Development | Application Development & Maintenance | Best Buy | A5-071 | Ph.: (612) 291-2350 | Cell: (612) 231-0011   [cid:image001.jpg@01CCB595.50739B50] <http://www.bestbuy.com/>

TFS Requests:  Service Now <https://bestbuy.service-now.com/> – Self Service -> Request Catalog -> IT Infrastructure -> TFS Support Requests

**Comment by chrismbarr** [comment]
@anthony-gregg I haven't changed the defaults, so yes I suppose so.  

I think I'm going to turn this off now since it's not something I need.
